Module: Reverse a single HI16/LO16 pair correctly

unknownbrackets opened the pull request:
If there is only one imported variable as a HI16/LO16, unloading the module wasn't properly reversing the link.  See #13104.

Have not actually tested this as fixing anything, but logically it needs to reset the HI16 chain if it unlinks the same thing it just linked.

-[Unknown]

unknownbrackets committed:
Module: Reverse a single HI16/LO16 pair correctly.

If there is only one imported variable as a HI16/LO16, unloading the
module wasn't properly reversing the link.  See #13104.

hrydgard commented on Core/HLE/sceKernelModule.cpp:
```diff
@@ -557,17 +557,17 @@ void __KernelModuleShutdown()
+static void WriteVarSymbol(u32 exportAddress, u32 relocAddress, u8 type, bool reverse = false) {
 	// We have to post-process the HI16 part, since it might be +1 or not depending on the LO16 value.
 	static u32 lastHI16ExportAddress = 0;
+	// This is just a list of recent HI16s for the same export address.
 	static std::vector<HI16RelocInfo> lastHI16Relocs;
 	static bool lastHI16Processed = true;
+	static bool lastHI16Reverse = false;
```
I'm wondering if these shouldn't belong to the module instead of being static (effectively global)? Or maybe it doesn't matter?

unknownbrackets replied:
It can't be interleaved.  These values shouldn't even last for the entire lifetime of the module.

Supposing a module importing 20 variables, each using the common case of a single HI16/LO16 pair.  In that case, the `lastHI16Relocs` vector would never hold more than one item, and would be cleared 20 times on link within  `if (lastHI16ExportAddress != exportAddress || lastHI16Reverse != reverse) {`.  It would be cleared another 20 times on unlink.

It's not some long lived list of relocations.  It's just for the HI16s that need to be adjusted for the next LO16s.

-[Unknown]

unknownbrackets replied:
I moved it to a state object - I suppose if two modules both imported exactly the same variable there might be a similar problem?  But hopefully this makes it clearer to everyone that this isn't some module state.

-[Unknown]

hrydgard replied:
That's much better! I avoid writing code with statics like that these days, prefer to let the type/lifetime system of the language enforce proper usage...

unknownbrackets committed:
Module: Keep HI16/LO16 in a temp state object.

This doesn't need to live any longer than the link or unlink, so let's
just make that abundantly clear.

ANR2ME commented:
I think this is causing Ratchet & Clank Size Matters to get worse, multiplayer profile can't be Selected/Created anymore, thus can't progress further to test whether graphical glitch still exist or not.
